### ISSUE TITLE
More Data-driven Engineering for IOx

### ISSUE THREAD

**Issue by crepererum** [influxdata/influxdb_iox#5464]
# Why

During the IOx development, product scoping and operations, we face the following questions/challenges:

* Did change X improve/regress performance?
* https://github.com/influxdata/conductor/issues/958
* A/B testing for features / algorithms
* Many questions that people seem to wish to extract from query logs (that however are NOT persisted), e.g. https://github.com/influxdata/influxdb_iox/issues/4558

All these questions should be answered based on data. Ideally, this data shall be collected in neutral form (e.g. a CSV file) and then be analyzed with the same methods that we expect from proper medical studies instead of eye-balling[^1]. During this analysis, we want to have the ability to:

* Detect (and potentially remove) outliers
* Trim data ranges (e.g. to “X minutes after process start”)
* Remove / weight different samples (e.g. “their super bizarre query they run every morning”)

The data that we currently have is not suitable for this task, as the following subsections explain.

## Logs
We have logs, but they suffer from multiple drawbacks:

* **Correlation:** A single query has multiple log lines with the important information we need. Correlating / collecting them is cumbersome. This point could be worked around by emitting a single log line per relevant event with all metrics attached.
* **Retention:** Graylog has a very short retention period, so we would need to regularly (and automatically) extract the relevant data for long-time analysis.
* **Extraction:** Graylog is slow and its API rather average.

## Metrics (Prometheus / Grafana)
IOx also emits Prometheus-compatible metrics. However, they are limited as well:

* **Marginal distributions:** All metrics are marginal distributions. A detailed selection/filtering requires many labels with potentially larger cardinality – which is something Prometheus does not cope well with.
* **Binning:** [Histograms](https://prometheus.io/docs/concepts/metric_types/#histogram) (and [summaries](https://prometheus.io/docs/concepts/metric_types/#summary), but we do not use these) are discretized. This requires extreme care during analysis because it easily introduces biases (in mean and variance) and makes the estimation of small effects way harder. Also, our histograms make assumptions about the distribution of the data (= something exponential) to easily capture large ranges of data with minimal operation effort (= tuning the bin ranges).
* **Data Model:** Prometheus data model is based on time series data and while we can squeeze everything into a time series, the extraction of data is rather cumbersome.

## Traces
Traces in prod are so heavily sampled that they are barely existent and when they are, you may easily miss half of the trace.

## Binlog
The binlog contains a dump of the gRPC request and answer, but no internal metrics. Also, due to its nature (= contains request payload) its retention time is rather limited.

# What

This section mostly focuses on metrics about gRPC requests to the querier, but the creative reader can easily extrapolate this to other measurements.

An easy-to-use API within IOx to create one “entry” per gRPC request that contains:

* Start timestamp
* Duration
* IOx version as GIT hash
* Process UUID (i.e. a UUID created when the IOx process starts)
* Process uptime
* gRPC route/method
* (potentially) gRPC query parameters

Ideally, we would have a single “builder”/”handle” that we pass around and where different parts of the stack could easily attach fields / tags[^2]. The data is then collected in some central instance.

# How

The [“What” section](#what) already describes an API, but now we face the question of how to get the data from our cluster into someone's laptop. First, a quick feasibility check:

* Our current prod system has – according to [this query](https://grafana.prod01-eu-central-1.aws.influxdata.io/explore?orgId=1&left=%5B%22now-7d%22,%22now%22,%22Prometheus%20IOX%22,%7B%22refId%22:%22A%22,%22instant%22:true,%22range%22:true,%22exemplar%22:true,%22expr%22:%22sum%20(rate(grpc_requests_total%7Bjob%3D~%5C%22.*querier.*%5C%22%7D%5B10m%5D))%22%7D%5D) – at most 200 req/s (and this includes the uninteresting `Offset` requests).
* We assume an entry has 10 columns (timestamp + fields + tags)
* We assume that on the long-run, a reasonable data model would require not more than 256 bit per cell (128 bit for the column name-tag, 128 bit for the payload)
* We assume that we want to keep at max 1y of data only for a reasonable subset of our prod workload
* That results in: 
  $200 \frac{req}{s} \cdot 3600 \frac{s}{d} \cdot 24 \frac{s}{d} \cdot 365 \frac{d}{y} \cdot 10 \frac{column}{req} \cdot 256 \frac{bit}{column} \approx 1.9 \frac{Tbyte}{y}$ 
* Given that this a rather conservative measure, this is feasible.

Here are some ideas for the data collection:

## Single Log Line
Collect the metrics for the “event”/”entry” and create a single log line. All drawbacks described in the [“Logs” section](#logs) remain.

## Redis
Use redis to dump some events as structured data (potentially with some compression to keep the amount low).

## OpenTelemetry
We could use some model from the [OpenTelemetry](https://opentelemetry.io/) collection, e.g. [Exemplars](https://opentelemetry.io/docs/reference/specification/metrics/datamodel/#exemplars). However, it is unclear where to store the data and how to extract it for further analysis – we could (ab)use the [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/).

## IOx
Use the InfluxDB write protocol to store our own data in our own cluster. This would allow us to quickly filter+aggregate the data or download it (either by using the parquet files directly or by using the Flight client to download “query results”). This would also be a great opportunity to dogfood IOx.

[^1]: The author (@crepererum) admits that they also repeatedly forged a needle in the haystack by eye-balling different metrics / filters / accumulations until there was one configuration that supported an argument / story.

[^2]: The naming here has a reason as you will discover in the [“How” section](#how).

**Comment by alamb** [comment]
I wonder if this ticket is still in progress / tracking something actionable

**Comment by crepererum** [comment]
Yes, please keep this open. I have plans for this, but I need to shape them. This will heavily lean on the work that @adrian-thurston is doing. I'll sit down next week and write down a vision.
